src/schema-validator.ts

```diff
@@ -1,82 +1,34 @@
 
-const getLineFromOffset = (lines: number[], offset: number): number => {
-  for (let index = 1; index < lines.length; index++) {
-    const newLineOffset = lines[index]
-    if (newLineOffset > offset) {
-      return index
-    }
-  }
-  return lines.length
-}
-
-const getLineFromInstancePath = (instancePath: string, cst: CSTRepresentation) => {
-  const pathItems = instancePath.split('/').slice(1)
-
-  // If there is no path no line can be found
-  if (pathItems.length === 0) return undefined
-
-  const getLineFromDoc = (doc: Document): number | undefined => {
-    let line: number | undefined = undefined
-    CST.visit(doc, (item, path) => {
-      const currentPathValue = pathItems[path.length - 1]
-      const num = Number(currentPathValue)
-      // If path step is a number
-      if (isNaN(num)) {
-        if (!CST.isScalar(item.value)) return undefined
-
-        const key = item.key as SourceToken
-        if (!key) return undefined
-
-        // If note key is not the current path value skip this node and its childs and go to next sibling
-        if (key.source !== currentPathValue) {
-          return CST.visit.SKIP
-        }
-        if (path.length === pathItems.length) {
-          // If it has the same value, check if it is the last item in the path, if so the item is found and finish visit
-          const index = getLineFromOffset(cst.lines, key.offset)
-          line = index
-          return CST.visit.BREAK
-        }
-      } else {
-        if (num !== path[path.length - 1][1]) {
-          // If number in the instance path provided is not the same as the current traversal path skip to the next sibling
-          // I.e. instancePath: /files/1/source, num = 1, path = [['value', 0], ['value', 0]] <- This is not a match because 1 != 0
-          return CST.visit.SKIP
-        }
-
-        if (!CST.isScalar(item.value)) return undefined
-
-        if (path.length === pathItems.length) {
-          // If it has the same value, check if it is the last item in the path, if so the item is found and finish visit
-
-          const index = getLineFromOffset(cst.lines, item.value.offset)
-          if (index > 0) {
-            line = index
-            return CST.visit.BREAK
-          }
-        }
-      }
-      // Otherwise go on visiting
-      return undefined
-    })
-    return line
-  }
-
-  return cst.tokens.reduce<number | undefined>((acc, t) => getLineFromDoc(t as Document) ?? acc, undefined)
+const getLineFromInstancePath = (
+  instancePath: string,
+  parsedDocument: Document.Parsed<ParsedNode>,
+  lineCounter: LineCounter,
+): number | undefined => {
+  const paths = instancePath.split('/').filter(p => p.length > 0)
+
+  const optionalNode = O.fromNullable(parsedDocument.getIn(paths, true) as Scalar)
+
+  return pipe(
+    optionalNode,
+    O.map(node => O.fromNullable(node.srcToken?.offset)),
+    O.flatten,
+    O.map(n => lineCounter.linePos(n).line),
+    O.toUndefined,
+  )
```

Comment:
This can be simplified to: 

```suggestion
  const optionalNode = parsedDocument.getIn(paths, true) as Scalar
  const offset = optionalNode?.srcToken?.offset ?? -1
  if (offset !== -1) lineCounter.linePos(offset).line
  return undefined
```

WDYT? Then we can also remove the use of `fp-ts`'s `Optional` here, since optionals are built into the language 😄

Reply:
Since we want to go more functional I think I will keep the code, if you don't mind. Although there are nullable types on TS,  using `Optional` lets using `map` and other functional functions, so there is a chain of functions applied.

Reply:
My point was mainly that: 
- we use the language builtins
- the code is shorter/simpler (less intermediary steps) and therefore easier to reason about for the reader

But I'm fine with keeping it in 😄